lib/eth/types/faceted_numeric.rb

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2016-2022 The Ruby-Eth Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# -*- encoding : ascii-8bit -*-
+
+# Provides the {Eth} module.
+module Eth
+  
+  # Provides the `Types` module supporting various data types.
+  module Types
+
+    # Provides a faceted, numeric data type to help alleviate
+    #   the pain of converting from a keccak256 binary string -> hex -> prefixed hex,
+    #   vice versa, and other conversions.
+    #
+    # @see {https://github.com/q9f/secp256k1.cr/blob/master/src/secp256k1/num.cr} Num from secp256k1.cr
+    # @see #{Eth::Util} is used heavily throughout
+    # @example Pass in a 0x-prefixed hex string, and coerce to keccak256 binary
+    #   FacetedNumeric.new('0xDEADBeeF').to_binary
+    #   FacetedNumeric['0xDEADBeeF'].to_keccak256
+    # @raise [ArgumentError] raised if input is a String but
+    #   isn't hex or bytes, or if input is an Array but not
+    #   a list of ascii-8bit numbers
+    class FacetedNumeric < Numeric
```

Comment:
It is to be expected that this datatype is to be used across the entire library, I would encourage you to pick a short name for the class, e.g., `::Num`, and use the documentation to elaborate on the nature of this faceted numeric type.

```suggestion
    class Num < Numeric
```

Reply:
You think it should be a top level class? Get rid of `Types` namespace?